File: Leetcode/110/solution.py
# https://leetcode.com/problems/balanced-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    # Recursively determine if a binary tree is balanced
    # Visits each node once: O(n) time, O(h) space
    def isBalanced(self, root):
        # Find the height of a tree given its root
        # return -1 if unbalanced
        def height(root):
            # Base case
            if root is None:
                return 0

            left_height  = height(root.left)
            right_height = height(root.right)

            # If tree is unbalanced, return -1
            if left_height == -1 or right_height == -1 or \
                abs(left_height - right_height) > 1:
                return -1

            # Otherwise, return height of tree
            return 1 + max(left_height, right_height)

        return height(root) != -1

File: Leetcode/110/solution.java
// https://leetcode.com/problems/balanced-binary-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // Recursively determine if a binary tree is balanced
    // Visits each node once: O(n) time, O(h) space
    public boolean isBalanced(TreeNode root) {
        return height(root) != -1;
    }

    // Find the height of a tree given its root
    // return -1 if unbalanced.
    private int height(TreeNode root) {
        // Base case
        if (root == null) return 0;

        int leftHeight = height(root.left);
        int rightHeight = height(root.right);

        // If tree is unbalanced, return -1
        if (leftHeight == -1 || rightHeight == -1 ||
            Math.abs(leftHeight - rightHeight) > 1) return -1;

        // Otherwise, return height of tree
        return 1 + Math.max(leftHeight, rightHeight);
    }
}
